modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java

```diff
@@ -535,17 +535,53 @@ protected void unsetAssetListAssetEntryQueryProcessor(
 
 		//if (_assetListConfiguration.considerAllSegmentsForUser()) {
 		if (true) {
-			for (long segmentsEntryId : _getCombinedSegmentsEntryIds(segmentsEntryIds)) {
-				AssetEntryQuery assetEntryQuery = getAssetEntryQuery(
-					assetListEntry, segmentsEntryId, userId);
+			if ((start == QueryUtil.ALL_POS) && (end == QueryUtil.ALL_POS)) {
+				for (long segmentsEntryId :
+						_getCombinedSegmentsEntryIds(segmentsEntryIds)) {
 
-				List<AssetEntry> assetEntries = _search(
-					assetListEntry.getCompanyId(), assetEntryQuery);
+					AssetEntryQuery assetEntryQuery = getAssetEntryQuery(
+						assetListEntry, segmentsEntryId, userId);
 
-				dynamicAssetEntries.addAll(assetEntries);
+					List<AssetEntry> assetEntries = _search(
+						assetListEntry.getCompanyId(), assetEntryQuery);
+
+					dynamicAssetEntries.addAll(assetEntries);
+				}
 			}
+			else {
+				int correctedStart = start;
+				int correctedEnd = end;
+				int subtotal;
+
+				for (long segmentsEntryId :
+						_getCombinedSegmentsEntryIds(segmentsEntryIds)) {
+
+					AssetEntryQuery assetEntryQuery = getAssetEntryQuery(
+						assetListEntry, segmentsEntryId, userId);
+
+					List<AssetEntry> assetEntries = _search(
+						assetListEntry.getCompanyId(), assetEntryQuery);
+
+					int entriesSize = assetEntries.size();
 
-			dynamicAssetEntries.subList(start, end);
+					subtotal =+ entriesSize;
+
+					if (subtotal < start) {
+						correctedEnd =- entriesSize;
+						correctedStart =- entriesSize;
+
+						continue;
+					}
+
+					dynamicAssetEntries.addAll(assetEntries);
+
+					if (subtotal > end) {
+						break;
+					}
+				}
+
+				dynamicAssetEntries.subList(correctedStart, correctedEnd);
+			}
 		}
```

Comment:
Maybe we could study to implement something like this:

```
	int remaining = Math.max(0, end - start);
				int subtotal = 0;
				int count = 0;

				for (long segmentsEntryId :
						_getCombinedSegmentsEntryIds(segmentsEntryIds)) {

					count = getAssetEntriesCount(
						assetListEntry, new long[] {segmentsEntryId}, userId);

					if ((subtotal + count) < start) {
						subtotal += count;

						continue;
					}

					AssetEntryQuery assetEntryQuery = getAssetEntryQuery(
						assetListEntry, segmentsEntryId, userId);

					List<AssetEntry> assetEntries = _search(
						assetListEntry.getCompanyId(), assetEntryQuery);

					count = assetEntries.size();

					List<AssetEntry> assetEntriesSubList = assetEntries.subList(
						Math.max(start - subtotal, 0),
						Math.min(remaining, count));

					dynamicAssetEntries.addAll(assetEntriesSubList);

					subtotal += count;
					remaining -= assetEntriesSubList.size();

					if (remaining < 0) {
						break;
					}
				}
```
The number of queries (database or index queries) is the same, but we avoid, as much as possible, loading records in memory.